src/h264rtpdepacketizer.cpp

```diff
@@ -0,0 +1,138 @@
+                                               message_vector::iterator last) {
+	message_vector out = {};
+	auto fua_buffer = std::vector<std::byte>{};
+
+	for (auto it = first; (it - 1) != last; it++) {
+		auto pkt = it->get();
+		auto pktParsed = reinterpret_cast<const rtc::RtpHeader *>(pkt->data());
+		auto headerSize =
+		    rtpHeaderSize + pktParsed->csrcCount() + pktParsed->getExtensionHeaderSize();
+		auto naluType = pkt->at(headerSize) & naluTypeBitmask;
+
+		if (fua_buffer.size() != 0 || naluType == naluTypeFUA) {
+			if (fua_buffer.size() == 0) {
+				fua_buffer.push_back(std::byte(0));
+			}
+
+			std::copy(pkt->begin() + headerSize + fuaHeaderSize, pkt->end(),
+			          std::back_inserter(fua_buffer));
+
+			if ((pkt->at(headerSize + 1) & fuaEndBitmask) != std::byte(0)) {
+				auto naluRefIdc = pkt->at(headerSize) & naluRefIdcBitmask;
+				auto fragmentedNaluType = pkt->at(headerSize + 1) & std::byte(naluTypeBitmask);
+				fua_buffer.at(0) = naluRefIdc | fragmentedNaluType;
+
+				out.push_back(make_message(fua_buffer));
+				fua_buffer.clear();
+			}
+		} else if (naluType > std::byte(0) && naluType < std::byte(24)) {
+			out.push_back(make_message(pkt->begin() + headerSize, pkt->end()));
+		} else if (naluType == naluTypeSTAPA) {
+			auto currOffset = stapaHeaderSize + headerSize;
+
+			while (currOffset < pkt->size()) {
+				auto naluSize =
+				    uint16_t(pkt->at(currOffset)) << 8 | uint8_t(pkt->at(currOffset + 1));
+
+				currOffset += 2;
+
+				if (pkt->size() < currOffset + naluSize) {
+					throw std::runtime_error("STAP-A declared size is larger then buffer");
+				}
+
+				out.push_back(
+				    make_message(pkt->begin() + currOffset, pkt->begin() + currOffset + naluSize));
+				currOffset += naluSize;
+			}
+
+		} else {
+			throw std::runtime_error("Unknown H264 RTP Packetization");
+		}
+	}
+
+	return out;
+}
+
+void H264RtpDepacketizer::incoming(message_vector &messages, const message_callback &) {
+	for (auto message : messages) {
+		if (message->type != Message::Binary) {
+			continue;
+		}
+
+		if (message->size() < sizeof(RtpHeader)) {
+			PLOG_VERBOSE << "RTP packet is too small, size=" << message->size();
+			continue;
+		}
+
+		this->rtp_buffer.push_back(make_message(message->begin(), message->end()));
+	}
+
+	while (this->rtp_buffer.size() != 0) {
+		uint32_t current_timestamp = 0;
+		size_t packets_in_timestamp = 0;
+
+		for (const auto &pkt : this->rtp_buffer) {
+			auto p = reinterpret_cast<const rtc::RtpHeader *>(pkt->data());
+
+			if (current_timestamp == 0) {
+				current_timestamp = p->timestamp();
+			} else if (current_timestamp != p->timestamp()) {
+				break;
+			}
+
+			packets_in_timestamp++;
+		}
+
+		if (packets_in_timestamp == this->rtp_buffer.size()) {
+			break;
+		}
+
+		auto first = this->rtp_buffer.begin();
+		auto last = this->rtp_buffer.begin() + (packets_in_timestamp - 1);
+
+		messages = buildFrame(first, last);
```

Comment:
I think there are issues with the handling of `messages`. For instance:
- If there is a single timestamp in `rtp_buffer` (for instance a single frame), no new frame is depacketized (because of the break just above), in that case it looks like input messages in `messages` won't be cleared and will leak to the next element in the media processing chain.
- If there are two frames depacketized in a single call, `messages` will be replaced for each frame, so frames will be dropped and only the last one will be passed to the chain.

Reply:
Maybe I'm missing parts of the logic but if the principle is to flush the current frame when the next timestamp is seen, couldn't such a simple approach do the job for `H264RtpDepacketizer::incoming`?
```cpp
message_vector result;
for (auto message : messages) {
    [...] // check message type and size
    auto p = reinterpret_cast<const RtpHeader *>(message->data());
    if (!rtp_buffer.empty() && current_timestamp != p->timestamp()) {
        result.push_back(buildFrame(rtp_buffer.begin(), rtp_buffer.end()));
        rtp.buffer.clear();
   }
   current_timestamp = p->timestamp();
   rtp_buffer.push_back(std::move(message));
}
messages.swap(result);
```
`current_timestamp` could be a class member (or read from a packet in `rtp_buffer` before the loop).

Reply:
Both should be handled now!

> If incoming RTP packets aren't enough to build a frame
`messages.clear()` is called so messages aren't leaked

> two frames in a singe call
I merge the lists now. If a incoming RTP packet results in multiple frames being available it works!

Reply:
It looks correct now, even if I'm still a bit puzzled by the convoluted approach.